src/main/java/com/yahoo/bullet/record/BulletRecord.java

```diff
@@ -5,222 +5,84 @@
-import java.io.IOException;
-import java.io.ObjectInputStream;
-import java.io.ObjectOutputStream;
 import java.io.Serializable;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
-import java.util.Objects;
 
 /**
- * This wraps the data that should be sent to Bullet. It is a {@link Serializable} object.
- * It also implements {@link Iterable} and can be used in for-each loops. The various
- * set methods can be used to insert fields into the record. Field names are to be provided as Strings and must be
- * unique, otherwise the duplicate takes precedence over the first.
+ * Data sent to bullet should be wrapped in a class that extends this abstract class. It is
+ * {@link Serializable} and {@link Iterable} so records can be used in for-each loops. The various
+ * set methods can be used to insert fields into the record. They are implemented by default to use
+ * the base {@link #set(String, Object)} method which should be overridden in child classes. Field
+ * names are to be provided as Strings and must be unique, otherwise the duplicate takes precedence
+ * over the first by default.
  *
  * When inserting into the Record, methods are explicitly provided for each type supported (listed
  * below). When reading from the Record, which may not be something needed by users of this very
  * frequently (as it is the entry point into Bullet), a generic Object is returned instead. Casting
  * it is left to the user.
  *
- * By default, when reading, the record will not present the data as valid Java objects. It will
- * keep it serialized till any of the get/set methods are called. This makes the object cheap to
- * send through repeated read-write cycles without modifications. You can force a read by either
- * calling a get/set method or using {@link #forceReadData()}.
- *
  * The record currently supports these fields:
  * <pre>
- * Primitives: Boolean, Long, Double, String
+ * Primitives: Boolean, Integer, Long, Float, Double, String
  * Complex: {@code Map <String, "Primitives">, Map<String, Map<String, "Primitives">>, List<Map<String, "Primitives">},
  *           where "Primitives" refers to the afore-mentioned Primitives.
  * </pre>
  *
  * Instead of setting a field to null (you cannot for top level Java primitives), avoid setting it instead.
  */
-@Slf4j @Setter(AccessLevel.PACKAGE) @NoArgsConstructor
-public class BulletRecord implements Iterable<Map.Entry<String, Object>>, Serializable {
-    public static final long serialVersionUID = 926415013785021742L;
-
-    private boolean isDeserialized = true;
-    private byte[] serializedData;
-    private Map<String, Object> data = new HashMap<>();
-    private static final SpecificDatumWriter<BulletAvro> WRITER = new SpecificDatumWriter<>(BulletAvro.class);
+public abstract class BulletRecord implements Iterable<Map.Entry<String, Object>>, Serializable {
```

Comment:
This should not be an Iterable of Map.Entry. We don't want other implementation of this to have to wrap their data in a Map first (you can't create Map.Entry). You can use Pair from apache if you like or another class that we implement here (Field { String name, Object value })?